core/trino-main/src/main/java/io/trino/operator/output/FlatteningPositionsAppender.java

```diff
@@ -0,0 +1,119 @@
+ * The {@link Block} is flattened if necessary so that the {@link #delegate} {@link PositionsAppender#append(IntArrayList, Block)}
+ * always gets flat {@link Block} and {@link PositionsAppender#appendRle(RunLengthEncodedBlock)} always gets {@link RunLengthEncodedBlock}
+ * with {@link RunLengthEncodedBlock#getValue()} being flat {@link Block}.
+ */
+public class FlatteningPositionsAppender
+        implements PositionsAppender
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(FlatteningPositionsAppender.class).instanceSize();
+
+    private final PositionsAppender delegate;
+
+    public FlatteningPositionsAppender(PositionsAppender delegate)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    @Override
+    public void append(IntArrayList positions, Block source)
+    {
+        if (source instanceof RunLengthEncodedBlock) {
+            delegate.appendRle(flatten((RunLengthEncodedBlock) source, positions.size()));
+        }
+        else if (source instanceof DictionaryBlock) {
+            appendDictionary(positions, (DictionaryBlock) source);
+        }
+        else {
+            delegate.append(positions, source);
+        }
+    }
+
+    @Override
+    public void appendRle(RunLengthEncodedBlock source)
+    {
+        if (source.getPositionCount() == 0) {
+            return;
+        }
+        delegate.appendRle(flatten(source, source.getPositionCount()));
+    }
+
+    @Override
+    public Block build()
+    {
+        return delegate.build();
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE + delegate.getRetainedSizeInBytes();
+    }
+
+    @Override
+    public long getSizeInBytes()
+    {
+        return delegate.getSizeInBytes();
+    }
+
+    private void appendDictionary(IntArrayList positions, DictionaryBlock source)
+    {
+        Block dictionary = source.getDictionary();
+
+        while (dictionary instanceof RunLengthEncodedBlock || dictionary instanceof DictionaryBlock) {
+            if (dictionary instanceof RunLengthEncodedBlock) {
+                // if at some level dictionary contains only a single value then it can be flattened to rle
+                appendRle(new RunLengthEncodedBlock(((RunLengthEncodedBlock) dictionary).getValue(), positions.size()));
+                return;
+            }
+
+            // dictionary is a nested dictionary. we need to remap the ids
+            DictionaryBlock nestedDictionary = (DictionaryBlock) dictionary;
+            positions = mapPositions(positions, source);
+            dictionary = nestedDictionary.getDictionary();
+            source = nestedDictionary;
+        }
+        delegate.append(mapPositions(positions, source), dictionary);
+    }
+
+    private RunLengthEncodedBlock flatten(RunLengthEncodedBlock source, int positionCount)
+    {
+        Block value = source.getValue().getSingleValueBlock(0);
```

Comment:
I would still add positions == 0 check here. Depending on `RunLengthEncodedBlock` internals to always have single position seems fragile